trio/_core/_io_windows.py

```diff
@@ -97,219 +222,280 @@ class CompletionKeyEventInfo:
-                    transferred = entry.dwNumberOfBytesTransferred
-                    info = CompletionKeyEventInfo(
-                        lpOverlapped=overlapped,
-                        dwNumberOfBytesTransferred=transferred,
-                    )
-                    queue.put_nowait(info)
-
-    def _iocp_thread_fn(self):
-        # This thread sits calling GetQueuedCompletionStatusEx forever. To
-        # signal that it should shut down, the main thread just closes the
-        # IOCP, which causes GetQueuedCompletionStatusEx to return with an
-        # error:
-        IOCP_CLOSED_ERRORS = {
-            # If the IOCP is closed while we're blocked in
-            # GetQueuedCompletionStatusEx, then we get this error:
-            ErrorCodes.ERROR_ABANDONED_WAIT_0,
-            # If the IOCP is already closed when we initiate a
-            # GetQueuedCompletionStatusEx, then we get this error:
-            ErrorCodes.ERROR_INVALID_HANDLE,
-        }
-        while True:
-            max_events = 1
-            batch = ffi.new("OVERLAPPED_ENTRY[]", max_events)
-            received = ffi.new("PULONG")
-            # https://msdn.microsoft.com/en-us/library/windows/desktop/aa364988(v=vs.85).aspx
             try:
                 _check(
-                    kernel32.GetQueuedCompletionStatusEx(
-                        self._iocp, batch, max_events, received, 0xffffffff, 0
+                    kernel32.DeviceIoControl(
+                        self._afd,
+                        IoControlCodes.IOCTL_AFD_POLL,
+                        poll_info,
+                        ffi.sizeof("AFD_POLL_INFO"),
+                        poll_info,
+                        ffi.sizeof("AFD_POLL_INFO"),
+                        ffi.NULL,
+                        lpOverlapped,
                     )
                 )
             except OSError as exc:
-                if exc.winerror in IOCP_CLOSED_ERRORS:
-                    # The IOCP handle was closed; time to shut down.
-                    return
-                else:
-                    self._iocp_queue.append(exc)
-                    return
-            self._iocp_queue.append((batch, received[0]))
-            self._main_thread_waker.wakeup_thread_and_signal_safe()
+                if exc.winerror != ErrorCodes.ERROR_IO_PENDING:  # pragma: no cover
+                    raise
+
+            op = AFDPollOp(lpOverlapped, poll_info, waiters)
+            waiters.current_op = op
+            self._afd_ops[lpOverlapped] = op
+
+    async def _afd_poll(self, sock, mode):
+        base_handle = _get_base_socket(sock)
+        waiters = self._afd_waiters.get(base_handle)
+        if waiters is None:
+            waiters = AFDWaiters()
+            self._afd_waiters[base_handle] = waiters
+        if getattr(waiters, mode) is not None:
+            raise _core.BusyResourceError
+        setattr(waiters, mode, _core.current_task())
+        self._refresh_afd(base_handle)
```

Comment:
If _refresh_afd raises for some reason, you're definitely not going to be in a good state. Do you think it's worth trying to recover? I was worried initially that there might be an issue if `sock` is invalid, but I think `_get_base_socket` will raise in that case.

Reply:
We also call `_refresh_afd` from abort function context, and `handle_io` context, neither of which are the slightest bit recoverable. And we don't know any way that it could fail, I think?

....well, hmm. I guess in those two contexts we actually don't revalidate the handle before calling it. I guess the only way you could reach those points with an invalid handle is if you manually close a socket without calling `notify_closing`, so arguably we could say "not our problem". But I wonder if I can construct a case where it happens. I guess you need to do something like:

1. Task A calls `wait_readable`
2. Task B calls `wait_writable`
3. Task C closes the socket without telling us
4. Task A gets cancelled while leaving task B still running

I guess that could happen by accident – accidentally forget to close a socket, hit control-C to get out... hmm.

Reply:
...It turns out that this is quite a rabbit-hole. This is already a big complicated PR, so let's defer worrying about this for a follow-up issue: #1272

Reply:
Though I should also say, for purposes of this PR: while we should follow up, merging this as-is won't be terrible. Our epoll backend is structured the same way, so merging this won't create any *new* issues, plus it means we know that no-one's ever hit the issues in practice.